packages/fx-core/src/core/FxCoreImplementV3.ts

```diff
@@ -426,8 +427,25 @@ export class FxCoreV3Implement {
     return ok(ctx?.envVars);
   }
 
-  @hooks([ErrorHandlerMW, ProjectMigratorMWV3])
   async phantomMigrationV3(inputs: Inputs): Promise<Result<Void, FxError>> {
+    // If the project is invalid or upgraded, the ProjectMigratorMWV3 will not take action.
+    // Check invaliad/upgraded project here before call ProjectMigratorMWV3
+    const projectPath = (inputs.projectPath as string) || "";
```

Comment:
I remember similar logic exists in other place. If it's true, can we find a way to reuse the common part?

Reply:
Here are the error conditions for the specific command upgrade not for middleware. 
In the ProjectMigratorMWV3, we do this check as well and the handle result is not the same. It will not return error but skip upgrade flow for these cases.
The common logic is wrapped as functions already.